cs3client/file.py

```diff
@@ -334,3 +357,261 @@ def list_dir(
         self._log.debug(f'msg="Invoked ListContainer" trace="{res.status.trace}"')
         for info in res.infos:
             yield info
+
+    def _set_lock_using_xattr(self, auth_token, resource: Resource, app_name: str, lock_id: int | str) -> None:
+        """"
+        Set a lock to a resource with the given value metadata and appname as holder
+
+        :param resource: Resource to set lock to
+        :param app_name: Application name
+        :param lock_id: Metadata lock value
+        :return: None (Success)
+        :raises: NotFoundException (File not found)
+        :raises: AuthenticationException (Authentication Failed)
+        :raises: UnknownException (Unknown Error)
+        """
+        self._log.debug(f'msg="Using xattrs to execute SetLock" {resource.get_file_ref_str()}" value="{lock_id}"')
+        # check if the lock is already set
+        try:
+            _ = self.stat(auth_token, resource).arbitrary_metadata.metadata[LOCK_ATTR_KEY]
+        except KeyError:
+            expiration = int(time.time() + self._config.lock_expiration)
+            self.set_xattr(resource, LOCK_ATTR_KEY, f"{app_name}!{lock_id}!{expiration}", None)
+            return
+
+    def set_lock(self, auth_token: tuple, resource: Resource, app_name: str, lock_id: int | str) -> None:
+        """
+        Set a lock to a resource with the given value and appname as holder
+
+        :param auth_token: tuple in the form ('x-access-token', <token>) (see auth.get_token/auth.check_token)
+        :param resource: Resource to set lock to
+        :param app_name: Application name
+        :param lock_id: encoded lock_id or metadata lock value if using xattr
+        :return: None (Success)
+        :raises: NotFoundException (File not found)
+        :raises: AuthenticationException (Authentication Failed)
+        :raises: UnknownException (Unknown Error)
+        """
+        # fallback to xattr if the storage does not support locks
+        if self._config.lock_by_setting_attr and self._config.lock_not_impl:
```

Comment:
Why is this an "AND"? If we want it to be configurable, then should be an OR. If we don't want it to be configurable, then we don't need `lock_by_setting_attr`.

Reply:
I misunderstood the logic. As agreed, there will be a comment explaining it.